### DISCUSSION TITLE
What happens during graceful shutdown?

### DISCUSSION THREAD

**Discussion by GrinningSoulGH** [hyperium/hyper#2515] [Q&A] [answered]
Graceful shutdown fails when I keep an open websocket connection (made with tokio_tungstenite's connect_async) in my request handler.

**Comment by seanmonstar** [comment, marked as answer]
Graceful shutdown will _try_ to signal shutdown intent based on what's available on the protocol:

- In HTTP/1, it will close the connection one the next response has been written. If the connection is currently idle, waiting for a new request, it will just close instantly.
- In HTTP/2, it will start the graceful shutdown dance by first sending a GOAWAY with the last stream set to the max. Then, after a round-trip-time, it will cap the last stream to whatever has been received so far, sending a new GOAWAY. It will then wait for all currently open streams (multiplexed requests on the same connection) to close by returning a response.

Notably, it cannot trigger shutdowns in protocols hyper doesn't understand. Connections that have been upgraded are no longer monitored by hyper, and shutdown in those protocols must be handled separately. It also won't _force_ a shutdown. To do that, you'll need to include a timeout after starting to trigger the graceful shutdown.
